storage/src/adb/current.rs

```diff
@@ -195,54 +204,88 @@ impl<
     /// Updates `key` to have value `value`. If the key already has this same value, then this is a
     /// no-op. The operation is reflected in the snapshot, but will be subject to rollback until the
     /// next successful `commit`.
-    pub async fn update(
-        &mut self,
-        hasher: &mut H,
-        key: K,
-        value: V,
-    ) -> Result<UpdateResult, Error> {
-        let update_result = self.any.update(hasher, key, value).await?;
+    pub async fn update(&mut self, key: K, value: V) -> Result<UpdateResult, Error> {
+        let update_result = self.any.update(key, value).await?;
+        self.hasher.graft(self.any.ops.take().unwrap());
```

Comment:
It feels like a lot of this complexity comes from having a long-lived `hasher` instance. If it only lived for the duration of `graft`, I suspect you could use references instead of this `Option<>` + `take` approach (which may simplify a lot)?

Reply:
With this approach, I'm not sure how you could concurrently prove/verify over immutable data (which seems like a nice feature to provide).

Reply:
got it working with the long lived hasher